<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <meta http-equiv="X-UA-Compatible" content="ie=edge" />
        <title>Document</title>
    </head>
    <body>
        <script>
            const promise = new Promise((resolve, reject) => {
                reject(22);
            });
            console.log('promise: ', promise);

            // promise.then(data => console.log(data));
            // promise.catch(data => console.log(data));
            // console.log('promise: ', promise);

            // const p1 = new Promise(function(resolve, reject) {
            //     setTimeout(() => reject(new Error('fail')), 4000);
            // });

            // const p2 = new Promise(function(resolve, reject) {
            //     setTimeout(() => resolve(p1), 2000);
            // });

            // p1.then(console.log, console.error);

            // p2.then(result => console.log('p2', result)).catch(error =>
            //     console.log('p2', error)
            // );

            // p.then(val => console.log('fulfilled:', val)).catch(err =>
            //     console.log('rejected', err)
            // );
            // // 等同于
            // p.then(val => console.log('fulfilled:', val)).then(null, err =>
            //     console.log('rejected:', err)
            // );

            // const promise = new Promise(function(resolve, reject) {
            //     resolve('ok');
            //     throw new Error('test');
            // });
            // promise
            //     .then(function(value) {
            //         console.log(value);
            //     })
            //     .catch(function(error) {
            //         console.log(error);
            //     });
            // // ok

            // const someAsyncThing = function() {
            //     return new Promise(function(resolve, reject) {
            //         // 下面一行会报错，因为x没有声明
            //         resolve(x + 2);
            //     });
            // };

            // someAsyncThing().then(function() {
            //     console.log('everything is great');
            // });

            // setTimeout(() => {
            //     console.log(123);
            // }, 2000);
            // // Uncaught (in promise) ReferenceError: x is not defined
            // // 123

            // Promise.prototype.finally = function(callback) {
            //     let P = this.constructor;
            //     return this.then(
            //         value => P.resolve(callback()).then(() => value),
            //         reason =>
            //             P.resolve(callback()).then(() => {
            //                 throw reason;
            //             })
            //     );
            // };
            // Promise.resolve(2).finally(v => console.log(v));

            // // 简略实现ajax
            // function ajax(url, type) {
            //     return new Promise((resolve, reject) => {
            //         var XHR = new XHRHttpRequest();
            //         XHR.open(type, url, true);
            //         XHR.send();
            //         XHR.onreadystatechange = () => {
            //             if (XHR.readyState == 4) {
            //                 if (XHR.status == 200) {
            //                     try {
            //                         let response = JSON.parse(XHR.responseText);
            //                         resolve(response);
            //                     } catch (e) {
            //                         reject(e);
            //                     }
            //                 } else {
            //                     reject(new Error(XHR.statusText));
            //                 }
            //             }
            //         };
            //     });
            // }
            // const p1 = new Promise((resolve, reject) => {
            //     resolve('hello');
            // })
            //     .then(result => result)
            //     .catch(e => e);
            // const p2 = new Promise((resolve, reject) => {
            //     throw new Error('报错了');
            // })
            //     .then(result => result)
            //     .catch(e => e);
            // Promise.all([p1, p2])
            //     .then(result => console.log(result))
            //     .catch(e => console.log(e));
            // // ["hello", Error: 报错了]

            // const jsPromise = Promise.resolve('dfdsfsdf');
            // console.log('jsPromise: ', jsPromise);
        </script>
    </body>
</html>
